### PR TITLE
distinguish github_actions_labels for CPU vs. CUDA builds

### PR DESCRIPTION
**Pull request by h-vetinari**
Some feedstocks have moved to the cirun open-gpu server, and even fewer are using GPU agents (basically tensorflow & pytorch). Those are the rarest resource (currently also at halved capacity due to some issues with the physical GPUs in the server).

Additionally, we cannot yet separate build and test phases into different agents, c.f. https://github.com/conda-forge/pytorch-cpu-feedstock/issues/314 & https://github.com/conda-forge/conda-smithy/issues/1472

As a consequence, a given pytorch PR currently takes at least >24h to complete, and that's if there are no competing GPU builds. To alleviate the pressure on the rarest resource, I'd like to zip `github_actions_labels` into the `cuda_compiler` zip, because this would allow us to at least choose CPU agents for the non-CUDA builds. An example based on this PR can be found on this [branch](https://github.com/h-vetinari/pytorch-cpu-feedstock/tree/cpu_vs_gpu).

This would very likely break the rendering of existing feedstocks on the open-gpu server, but there are [currently](https://github.com/search?q=org%3Aconda-forge+github_actions_labels+path%3Arecipe%2Fconda_build_config.yaml&type=code) only 7 of those, out of which only 5 are cuda-enabled (and it won't concern CPU-only feedstocks due to being gated on `CF_CUDA_ENABLED`), so IMO this should be manageable.

@conda-forge/pytorch-cpu 
@conda-forge/tensorflow 
@conda-forge/jaxlib
@conda-forge/libmagma
@conda-forge/flash-attn

### CODE REVIEW

File: recipe/conda_build_config.yaml [hunk 1]
```diff
@@ -172,13 +172,20 @@ docker_image:                                       # [os.environ.get("BUILD_PLA
...
+  - dummy                       # [linux and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]
+
 zip_keys:
   -                             # [unix]
     - c_compiler_version        # [unix]
     - cxx_compiler_version      # [unix]
     - fortran_compiler_version  # [unix]
     - cuda_compiler             # [linux and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]
     - cuda_compiler_version     # [linux and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]
+    - github_actions_labels     # [linux and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]
```
Comment (hmaarrfk) [inline recipe/conda_build_config.yaml:188]: these zip keys have become almost impossible to read.

Comment (h-vetinari) [reply recipe/conda_build_config.yaml:188]: Not sure what alternative you'd have in mind - unless we completely redesign how conda works (or stop supporting CUDA builds), this is irreducible complexity IMO

Comment (hmaarrfk) [reply recipe/conda_build_config.yaml:188]: I think we can start with: https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/6917

**Comment by hmaarrfk** [comment]
I would prefer that users of the GPU runners throttle themselves instead of adding extra complexity.

The GPU runners aren't necessary in many cases. They also don't really guarantee that the builds don't fail (due to flaky tests).


They can "throttle" by falling back to the CPU runners and rerendering.

**Comment by h-vetinari** [comment]
IMO flaky tests can be skipped, and I'd much rather have an extra entry in the zip rather than sacrifice testing fidelity. The ideal would be to build on CPU agents and only test the GPU bits on a GPU runner, but given that that's not yet possible and pytorch has chosen to test also the GPU paths, I think the best (=least bad) tradeoff would be what this PR proposes.

The CUDA 11.8 discussion is IMO unrelated - I'm not opposed but I think it should be had on its own merits.

**Comment by hmaarrfk** [comment]
> The CUDA 11.8 discussion is IMO unrelated - I'm not opposed but I think it should be had on its own merits.

Its partially related. We are doing all this work and jumping through hoops for cuda builds. Some of the old tricks need to go, to make it possible for us to think of new ones.

File: recipe/conda_build_config.yaml [hunk 2]
```diff
@@ -172,13 +172,20 @@ docker_image:                                       # [os.environ.get("BUILD_PLA
   - quay.io/condaforge/linux-anvil-aarch64:alma9    # [os.environ.get("CF_CUDA_ENABLED", "False") == "True" and os.environ.get("BUILD_PLATFORM") == "linux-aarch64" and os.environ.get("DEFAULT_LINUX_VERSION", "alma9") == "alma9"]
   - quay.io/condaforge/linux-anvil-ppc64le:alma9    # [os.environ.get("CF_CUDA_ENABLED", "False") == "True" and os.environ.get("BUILD_PLATFORM") == "linux-ppc64le" and os.environ.get("DEFAULT_LINUX_VERSION", "alma9") == "alma9"]
 
+# to be overridden per feedstock; part of zip so that CPU builds can choose different agents than GPU builds
+github_actions_labels:          # [linux and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]
+  - dummy                       # [linux and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]
+  - dummy                       # [linux and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]
+  - dummy                       # [linux and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]
+
```
Comment (carterbox) [inline recipe/conda_build_config.yaml:180]: There needs to be a better explanation in the comments here on how to use this feature especially if you're going to break existing non-gpu-runner feedstocks.

**Review by carterbox — CHANGES_REQUESTED** [review]

**Comment by carterbox** [comment]
The only two feedstocks using GPU runners are pytorch and magma. For magma, I never got around to enabling any tests that need GPUS, so I can move that feedstock to the cpu runners if that helps. Though, TBH, magma is rarely built anymore.

My understanding is that the problem you're trying to solve is that pytorch is competing with itself for GPU runners because *ALL* of the builds are assigned to GPU runners (even the non-GPU builds). Is that correct?

**Comment by h-vetinari** [comment]
> My understanding is that the problem you're trying to solve is that pytorch is competing with itself for GPU runners because _ALL_ of the builds are assigned to GPU runners (even the non-GPU builds). Is that correct?

Exactly.

File: recipe/conda_build_config.yaml (hunk 2 above)
Comment (h-vetinari) [reply recipe/conda_build_config.yaml:180]: It'd just mean that `github_actions_labels` needs to be set once for non-CUDA feedstocks (i.e. those without any `{{ compiler("cuda") }}`), or thrice for CUDA-enabled feedstocks (because those need to override the global pinning, which has configs for `{None, 11.8, 12.6}`)

**Comment by jakirkham** [comment]
Wonder if it is worth reconsidering `ccache` or similar ( https://github.com/conda-forge/conda-forge.github.io/issues/389 )

We had tried that before. However enough is different between then and now that we could do things that were not previously possible

For example we can store artifacts on CI. So using that for caching is possible (without needing to create that from scratch)

File: recipe/conda_build_config.yaml (hunk 2 above)
Comment (carterbox) [reply recipe/conda_build_config.yaml:180]: Like this?

```yaml
github_actions_labels:  # [linux]
- cirun-openstack-cpu-large  # [linux]
- cirun-openstack-gpu-large  # [linux]
- cirun-openstack-gpu-large  # [linux]
```

```yaml
github_actions_labels:  # [linux]
- cirun-openstack-cpu-large  # [linux]
```

Comment (carterbox) [reply recipe/conda_build_config.yaml:180]:
```suggestion
# to be overridden per feedstock; part of zip so that CPU builds can choose different agents than GPU builds
github_actions_labels:          # [linux and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]
  - cirun-openstack-cpu-placeholder                       # [linux and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]
  - cirun-openstack-gpu-placeholder                       # [linux and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]
  - cirun-openstack-gpu-placeholder                       # [linux and os.environ.get("CF_CUDA_ENABLED", "False") == "True"]

```

**Review by carterbox — APPROVED** [review]
I think this will work, but as you mentioned it will break any of the 7 existing feedstocks who use the openstack server for CUDA builds. They will need to update their conda_build_config.yaml to have three items in the github_actions_labels. Otherwise, they will get a rendering error about the length of the zip not being equal.

**Comment by isuruf** [comment]
Which feedstocks are you planning to use it for? If it's small, you could do it as a migrator.

**Comment by isuruf** [comment]
Also, is it possible to only set this in the feedstock `recipe/conda_build_config.yaml` and extend the zip_keys there?

**Comment by h-vetinari** [comment]
> Also, is it possible to only set this in the feedstock `recipe/conda_build_config.yaml` and extend the zip_keys there?

unfortunately not, because it's not possible to extend an existing zip_key (I tried). See the POC [branch](https://github.com/h-vetinari/pytorch-cpu-feedstock/tree/cpu_vs_gpu) that was rendered with the CBC from this PR.

> Which feedstocks are you planning to use it for? If it's small, you could do it as a migrator.

Primarily pytorch and tensorflow for now.

**Comment by isuruf** [comment]
We could just do
```yaml
skip: True   # [linux and (cuda_compiler_version == "None") == (github_actions_labels == "cirun-openstack-gpu-placeholder")]
```
after adding
```
github_actions_labels:                                      # [linux]
  - cirun-openstack-cpu-placeholder                 # [linux]
  - cirun-openstack-gpu-placeholder                 # [linux]
```
in the feedstock

**Comment by h-vetinari** [comment]
That seems pretty hacky to me (aside from the fact that smithy does not prune the jobs correctly - I [tested](https://github.com/h-vetinari/pytorch-cpu-feedstock/tree/cpu_vs_gpu_isuru) it).

I mean, if the resistance to this proposal is strong, I'm not gonna die on this hill, but to me it seems natural to add the runner-type (CPU vs. CUDA) to the same zip that decides whether it's a CPU or a CUDA build.

**Comment by hmaarrfk** [comment]
Looking at Pytorch v2.4.x https://github.com/conda-forge/pytorch-cpu-feedstock/pull/325
* The CPU jobs take about 3 hours to run. - 2 jobs
* The CUDA jobs take 9 hours - 4 jobs

So to save 6 / 42 hours we are adding complexity to anybody zipping keys at conda-forge. it just doesn't seem smart to me.

I'm not going to vote "for this approach", but I don't want to block it. Just I'm a little tired of fighting with these zipp'ed keys.

Lets keep it simple and not take precious human resources to discuss this.

IMO: just request the CPU runners on pytorch for things like the 2.4.x maintenance branch and one should be fine. We could also create ways to temporarily reduce the cuda target architectures during debugging sessions. These are "pytorch only" solutions that don't need to be imposed on the rest of conda-forge.

**Comment by h-vetinari** [comment]
### TL;DR: an extra zip entry costs ~nothing, but without we'd pay the extra runtime for every CI run

For pytorch main, the calculus is 3x CPU + 3xCUDA, and the aarch CPU build got longer because we're running parts of the test suite in emulation. So it's something like 12h runtime out of a total of 39, or almost a third. I don't care much what we do on the maintenance branches, but main should be relatively painfree to rebuild, and IMO we shouldn't accept an avoidable 50% more runtime.

The simple fact that the GPU agents are essentially our rarest and most precious CI resource would IMO be enough reason to do this - even if it were only for pytorch (which it isn't, because tensorflow will be in the same situation and has an even bigger matrix).

> we are adding complexity to anybody zipping keys at conda-forge.

This affects a grand total of 5 feedstocks in conda-forge, 2 of which would benefit from it. 🤷 

Overall, if this is so contentious, then I'll just go with Isuru's suggestion - but then don't start complaining about the superfluous CI workflows that this creates (half of which get skipped as soon as they hit conda-build) -- I find that quite a bit more painful than an extra zip entry, but _de gustibus non est disputandum_ I guess...

**Comment by h-vetinari** [comment]
> Just I'm a little tired of fighting with these zipp'ed keys.

FWIW, I'm happy to help out with issues on these things if you ping me. I haven't run into any unsolvable cases yet (well, aside from extending the zip locally, which conda-build doesn't support).

**Comment by h-vetinari** [comment]
Welp, given the concerns here I was planning to use Isuru's workaround, but on the pytorch feedstock that actually breaks windows
```
Checking out the ref
  "C:\Program Files\Git\bin\git.exe" checkout --progress --force -B main refs/remotes/origin/main
  Error: error: unable to create file .ci_support/linux_64_blas_implgenericc_compiler_version13channel_targetsconda-forge_maincuda_compilerNonecuda_compiler_versionNonecxx_compiler_version13github_actions_labelscirun-openstack-cpu-2xlargeis_rcFalse.yaml: Filename too long
  Error: error: unable to create file .ci_support/linux_64_blas_implgenericc_compiler_version13channel_targetsconda-forge_maincuda_compilerNonecuda_compiler_versionNonecxx_compiler_version13github_actions_labelscirun-openstack-gpu-2xlargeis_rcFalse.yaml: Filename too long
  [...]
```
Perhaps this is just a question of needing to add 
```
git config --system core.longpaths true
```
on the windows server? Any chance we could try to add that @baszalmstra @wolfv?

**Comment by hmaarrfk** [comment]
I don’t want to block this. Please disregard my concerns.

**Comment by isuruf** [comment]
We should probably reduce `250` at https://github.com/conda-forge/conda-smithy/blob/5228b1965e6c4ff2a4891aa238ff6311a15908b7/conda_smithy/configure_feedstock.py#L797

**Comment by h-vetinari** [comment]
> We should probably reduce `250` at https://github.com/conda-forge/conda-smithy/blob/5228b1965e6c4ff2a4891aa238ff6311a15908b7/conda_smithy/configure_feedstock.py#L797

Thanks for the proposition. I tested this in https://github.com/conda-forge/pytorch-cpu-feedstock/pull/332 but even with some more careful changes in https://github.com/conda-forge/conda-smithy/pull/2233, this either shortens the variant configs to the point that the key attributes (is it a CUDA build?) wouldn't be visible from the filename anymore, or still runs into issues with the filename lengths.

Despite the fact that the enthusiasm here is not exactly high, I think this is the right way to go about it conceptually (the zip that determines CUDA/non-CUDA is the canonical place for this distinction), and that the work-arounds are worse (filename length issues, aside from pointlessly doubling the number of linux jobs, etc.) than what they're meant to avoid (a single extra zip entry).

So I'll merge this and fix up the handful of affected feedstocks. The changes are minimal, e.g. I tested that this works for jaxlib:
```diff
--- a/recipe/conda_build_config.yaml
+++ b/recipe/conda_build_config.yaml
@@ -8,5 +8,8 @@ MACOSX_SDK_VERSION:        # [osx and x86_64]
 - '10.14'                  # [osx and x86_64]
 c_stdlib_version:          # [osx and x86_64]
 - '10.14'                  # [osx and x86_64]
-github_actions_labels:
-- cirun-openstack-cpu-large
+
+github_actions_labels:      # [linux]
+- cirun-openstack-cpu-large # [linux]
+- cirun-openstack-cpu-large # [linux]
+- cirun-openstack-cpu-large # [linux]
```

**Comment by isuruf** [comment]
Please don't merge things like this on your own without approval

**Comment by h-vetinari** [comment]
There were other approvals, and Mark withdrew his objections. You proposed alternatives but didn't request changes, much less provide arguments why this would be wrong. I'm willing to do the required clean-ups on the 5 affected feedstocks too, so this is a non-issue, unless you have a very concrete argument why this is the wrong approach.

**Comment by h-vetinari** [comment]
Also: I've tried all the suggested alternatives beforehand as well, and they were worse than what they were working around.

**Comment by isuruf** [comment]
I still Mark's objection. The zip_keys for docker_image and related ones are getting out of hand. I'd like to one day remove `docker_image` from that `zip_keys` and adding something like this would make it impossible to do down the road.

**Comment by h-vetinari** [comment]
> I'd like to one day remove `docker_image` from that `zip_keys` and adding something like this would make it impossible to do down the road.

I agree on this goal! And there's nothing standing in the way there AFAICT - as soon as we drop CUDA 11.8, we'll be able to drop `docker_image` from the zip; the image and the GHA label are not coupled (i.e. both non-CUDA and CUDA 12.6 use the alma9 image).

**Comment by isuruf** [comment]
Sure, but this `zip_key` is still getting out of hand, and I'd like to keep it as simple as possible.

**Comment by isuruf** [comment]
This is a conversation we need to have in an issue and I think the best course of action is to revert this PR, open an issue and if needed discuss in a core meeting.

**Comment by h-vetinari** [comment]
> Sure, but this `zip_key` is still getting out of hand, and I'd like to keep it as simple as possible.

We recently removed `c_stdlib_version` and `cdt_name`. So it's getting simpler already.

If we keep using the newest CUDA, we might even get to drop `{c,cxx,fortran}_compiler_version` from the zip. My point is that the `github_action_labels` fundamentally belongs in the same zip as `cuda_compiler_version`, and the results of working around this are worse than the impact it's trying to avoid.

**Comment by h-vetinari** [comment]
Wrote up an issue: https://github.com/conda-forge/conda-forge-pinning-feedstock/issues/6967

**Comment by isuruf** [comment]
Thank you!

**Comment by isuruf** [comment]
I started the revert PR at #6968
